accord-core/src/main/java/accord/messages/ReadData.java

```diff
@@ -53,7 +54,35 @@ public static ReadData create(TxnId txnId, Seekables<?, ?> scope, long executeAt
             return new ReadData(txnId, scope, executeAtEpoch, waitForEpoch);
         }
     }
+    private class ObsoleteTracker implements CommandListener
+    {
+        @Override
+        public void onChange(SafeCommandStore safeStore, SafeCommand safeCommand)
+        {
+            switch (safeCommand.current().status())
```

Comment:
This doesn't check whether the read already successfully completed before the transition to `PreApplied` or `Applied` for this command store. It may be that it wasn't obsoleted.

Reply:
good call... 

if we reply back, the listener still exists (as we didn't remove) so the next step would trigger an obsolete reply; which would get ignored...

We do the following on the `apply` failure case

```
node.commandStores().mapReduceConsume(this, waitingOn.stream(), forEach(in -> removeListener(in, txnId), node.agent()));
```

so could do the same in `ack`.

Now, the better solution is to do the same, but switch from `boolean` to `enum`.  if we reply back *before* `PreAccept`, then that's fine... so we should ignore in that case

ill look to improve this